src/dominion/player.rs

```diff
@@ -54,14 +74,213 @@ impl Player {
             self.shuffle_deck();
         }
 
+        // TODO handle empty deck
         self.hand.push(self.deck_pile.remove(0));
     }
 
     pub fn cleanup(&mut self) {
+        self.discard_pile.append(&mut self.in_play);
         self.discard_pile.append(&mut self.hand);
 
         for _ in 0..5 {
             self.draw_card();
         }
     }
+
+    pub fn start_turn(&mut self) -> Result<()> {
+        if let None = self.phase {
+            self.phase = Some(TurnPhase::Action {
+                action: 1,
+                buy: 0,
+                worth: 0,
+            });
+
+            Ok(())
+        } else {
+            Err(Error::WrongTurnPhase)
+        }
+    }
+
+    pub fn start_buy_phase(&mut self) -> Result<()> {
+        if let Some(TurnPhase::Action {
+            action: _,
+            mut buy,
+            worth,
+        }) = self.phase
+        {
+            self.phase = Some(TurnPhase::Buy { buy, worth });
+
+            Ok(())
+        } else {
+            Err(Error::WrongTurnPhase)
+        }
+    }
+
+    pub fn end_turn(&mut self) -> Result<()> {
+        if let Some(TurnPhase::Buy { .. }) = self.phase {
+            self.phase = None;
+
+            Ok(())
+        } else {
+            Err(Error::WrongTurnPhase)
+        }
+    }
+
+    pub fn play_card(&mut self, card_index: usize) -> Result<()> {
+        // TODO handle non-standard card actions
+        if card_index >= self.hand.len() {
+            Err(Error::InvalidCardIndex)
+        } else if let Some(phase) = &mut self.phase {
+            match phase {
+                TurnPhase::Action { action, buy, worth } => {
+                    if action > &mut 0 {
```

Comment:
Why is the `&mut` needed in front of the `0`?

Reply:
`action` is `&mut` so `0` needs to be the same type for comparison.

Reply:
Indeed. I must say that I am surprised that one cannot compare a non-`mut` to a `mut`, but OK, that's hoe it is. Would be interested in knowing the rational, however :)

But this seems to work: `*action > 0`, Not sure which variant is more "rusty", but to me this one feels more sane than creating a `mut` reference to an integer constant just to be able to do a comparison.

Reply:
@andebjor I'd have to agree. It is clearer to dereference action which is `Copy` and do the comparison.